Greentea netsocket more tests

VeijoPesonen opened the pull request:
### Description
Improving Greentea netsocket test case coverage
<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/guidelines.html#workflow (Pull request template)
-->


### Pull request type

[X] Test case
[ ] Fix  
[ ] Refactor  
[ ] New target  
[ ] Feature  
[ ] Breaking change

Commit:
Rely on ARM's and U-Blox's echo servers

Make the server as default. Use the hostnames instead of plain IP addresses.

VeijoPesonen commented on TESTS/netsocket/tcp_chargen_recv100k/main.cpp:
```diff
@@ -0,0 +1,205 @@
+    net->disconnect();
+}
+
+/** Generate RFC 864 example pattern.
```
Implementation provided by @SeppoTakalo

VeijoPesonen commented on tools/test_configs/ESP8266Interface.json:
```diff
@@ -13,7 +13,7 @@
         "echo-server-addr" : {
             "help" : "IP address of echo server",
-            "value" : "\"52.215.34.155\""
+            "value" : "\"echo.mbedcloudtesting.com\""
```
This won't work. Only IP address is accepted.

SeppoTakalo replied:
Where this limitation is?
Socket API is perfectly capable of accepting domain names.

VeijoPesonen replied:
Need to figure it out.

SeppoTakalo commented on TESTS/netsocket/tcp/tcpsocket_recv_100k.cpp:
```diff
@@ -0,0 +1,165 @@
+{
+    int err;
+    TCPSocket sock;
+    SocketAddress tcp_addr(MBED_CONF_APP_ECHO_SERVER_ADDR, 19);
```
SocketAddress does not resolve names. It is just a placeholder for IP addresses.
Use `NetworkInterface::gethostbyname()` or feed the string to `Socket` which translates its string through the `gethostbyname()`

SeppoTakalo replied:
So for example `TCPSocket::connect(const char *host, uint16_t port)` does the translation.

VeijoPesonen replied:
Ok, I just need to remember to make the same change to all netsocket test cases.

Commit:
Starts adding more TCP and UDP performance tests

Starts splitting Greentea test suites to TCP and UDP suites and adds
more test cases like

	tcpsocket_echotest_burst
	tcpsocket_echotest_burst_nonblock
	tcpsocket_endpoint_close
	tcpsocket_recv_100k
	tcpsocket_recv_100k
	tcpsocket_recv_timeout
	tcpsocket_send_timeout
	tcpsocket_thread_per_socket_safety
	udpsocket_echotest_burst
	udpsocket_echotest_burst_nonblock
	udpsocket_recv_timeout
	udpsocket_sendto_timeout

juhaylinen committed:
Update Greentea WiFi tests

Use similar network interface configuration as used in netsocket tests
Improve test coverage

Commit:
Adds more TCP and UDP test cases

Adds TCP and UDP test cases

        tcpsocket_open_limit
        udpsocket_open_limit

Commit:
Adds TCP and UDP test cases

	tcpsocket_open_close_repeat
	udpsocket_open_close_repeat

juhaylinen committed:
Update Greentea test configuration files

Add .json files for WicedInterface and SpwfSAInterface
Add target configurations for
MTB_UBLOX_ODIN_W2
MTB_ADV_WISE_1530
MTB_USI_WM_BN_BM_22
MTB_MXCHIP_EMW3166
NUCLEO_F401RE

Commit:
Adds TCP test case

                tcpsocket_connect_invalid

Commit:
Adds TCP test case

	tcpsocket_echotest
	tcpsocket_echotest_nonblock

Commit:
Removes obsolete TCP and UDP Greentea test cases

Commit:
Adds UDP Greentea test case

	udpsocket_sendto_invalid

Commit:
Adds TCP test case

	tcpsocket_send_repeat

Commit:
Adds UDP Greentea test case

        udpsocket_sendto_repeat

Commit:
Adds UDP Greentea test cases

	udpsocket_echotest
	udpsocket_echotest_nonblock

VeijoPesonen commented:
~~TCPSOCKET_RECV_100K_NONBLOCK and TCPSOCKET_RECV_100K are still unstable but people can start to review the other tests.~~

Wrong lwIP configuration (#define TCP_QUEUE_OOSEQ             0) was causing the issue

SeppoTakalo reviewed (changes requested):
Add error handling.

There are multiple cases where you call socket functions without testing the return code.
For example:
```
get_interface()->gethostbyname(...);
sock->open(...);
sock->connect(...);
sock->close();
```

Every socket call should be tested and compared against acceptable return values. Not only those that transfer the data, but also ones than translate names or create connection.

SeppoTakalo commented on TESTS/netsocket/tcp/main.cpp:
```diff
@@ -0,0 +1,115 @@
+utest::v1::status_t greentea_setup(const size_t number_of_cases)
+{
+    GREENTEA_SETUP(240, "default_auto");
+    _ifup();
```
Does the `TEST_ASSERT()` propagation work in `greentea_setup()`?
Are we able to see whether device did not join the network and therefore tests failed to run?

VeijoPesonen replied:
Yes, we are able to see if the problem arises from the handler. But there is still need to check the logs.

```
PS C:\ws\mbed-os> mbedhtrun -f .\BUILD\tests\K64F\ARM\TESTS\netsocket\udp\udp.bin -d D: -p COM14:9600
[1526317024.03][HTST][INF] host test executor ver. 1.2.0
[1526317024.03][HTST][INF] copy image onto target...
        1 file(s) copied.
[1526317032.47][MBED][WRN] Target ID not found: Skipping flash check and retry
[1526317032.47][HTST][INF] starting host test process...
[1526317033.09][CONN][INF] starting connection process...
[1526317033.09][CONN][INF] notify event queue about extra 60 sec timeout for serial port pooling
[1526317033.09][CONN][INF] initializing serial port listener...
[1526317033.09][SERI][INF] serial(port=COM14, baudrate=9600, read_timeout=0.01, write_timeout=5)
[1526317033.09][SERI][INF] reset device using 'default' plugin...
[1526317033.11][HTST][INF] setting timeout to: 60 sec
[1526317033.36][SERI][INF] waiting 1.00 sec after reset
[1526317034.36][SERI][INF] wait for it...
[1526317034.36][SERI][TXD] mbedmbedmbedmbedmbedmbedmbedmbedmbedmbed
[1526317034.36][CONN][INF] sending up to 2 __sync packets (specified with --sync=2)
[1526317034.36][CONN][INF] sending preamble '0ea1005b-0e2c-4dec-a15f-c947b6556639'
[1526317034.36][SERI][TXD] {{__sync;0ea1005b-0e2c-4dec-a15f-c947b6556639}}
[1526317034.49][CONN][RXD] mbedmbedmbedmbedmbedmbedmbedmbed
[1526317034.54][CONN][INF] found SYNC in stream: {{__sync;0ea1005b-0e2c-4dec-a15f-c947b6556639}} it is #0 sent, queued...
[1526317034.56][HTST][INF] sync KV found, uuid=0ea1005b-0e2c-4dec-a15f-c947b6556639, timestamp=1526317034.540000
[1526317034.57][CONN][INF] found KV pair in stream: {{__version;1.3.0}}, queued...
[[1526317034.59][HTST][INF] DUT greentea-client version: 1.3.0
1526317034.59][CONN][INF] found KV pair in stream: {{__timeout;120}}, queued...
[1526317034.59][HTST][INF] setting timeout to: 120 sec
[1526317034.62][CONN][INF] found KV pair in stream: {{__host_test_name;default_auto}}, queued...
[1526317034.62][HTST][INF] host test class: '<class 'mbed_host_tests.host_tests.default_auto.DefaultAuto'>'
[1526317034.62][HTST][INF] host test setup() call...
[1526317034.62][HTST][INF] CALLBACKs updated
[1526317034.62][HTST][INF] host test detected: default_auto
[1526317078.79][CONN][RXD] :46::FAIL: Expected 0 Was -3004
[1526317078.86][CONN][RXD] >>> failure with reason 'Assertion Failed' during 'Test Setup Handler'
[1526317078.90][CONN][INF] found KV pair in stream: {{max_heap_usage;0}}, queued...
[[1526317078.91][CONN][INF] found KV pair in stream: {{reserved_heap;0}}, queued...
1526317078.91][HTST][ERR] orphan event in main phase: {{max_heap_usage;0}}, timestamp=1526317078.897000
[1526317078.91][HTST][ERR] orphan event in main phase: {{reserved_heap;0}}, timestamp=1526317078.913000
[1526317078.93][CONN][INF] found KV pair in stream: {{end;failure}}, queued...
[1526317078.94][CONN][INF] found KV pair in stream: {{__exit;0}}, queued...
[1526317078.95][HTST][INF] __exit(0)
[1526317078.95][HTST][INF] __notify_complete(False)
[1526317078.95][HTST][INF] __exit_event_queue received
[1526317078.95][HTST][INF] test suite run finished after 44.36 sec...
[1526317078.96][CONN][INF] received special even '__host_test_finished' value='True', finishing
[1526317078.98][HTST][INF] CONN exited with code: 0
[1526317078.98][HTST][INF] Some events in queue
[1526317078.98][HTST][INF] stopped consuming events
[1526317078.98][HTST][INF] host test result() call skipped, received: False
[1526317078.98][HTST][INF] calling blocking teardown()
[1526317078.98][HTST][INF] teardown() finished
[1526317078.98][HTST][INF] {{result;failure}}
```

VeijoPesonen replied:
I'll add the required return value checks. ~~I wouldn't prefer not to because it will just add clutter on the test cases and assertions would reveal problems arising during setups.~~ Part of the test is to check that the desired behavior is the result of correct setup.

SeppoTakalo commented on TESTS/netsocket/udp/udpsocket_echotest.cpp:
```diff
@@ -0,0 +1,153 @@
+            sent = sock.sendto(udp_addr, tx_buffer, pkt_s);
+            if (sent != pkt_s) {
+                printf("[Round#%02d - Sender] error, returned %d\n", s_idx, sent);
+                continue;
```
Are you ignoring the `sendto()` failures?

VeijoPesonen replied:
Yes, I am because buffer comparison will reveal if data was not transmitted and received - retries are allowed. Checking for errors wouldn't make any difference.

cmonr commented:
@VeijoPesonen Was there a particular reason why this PR was give a new category?

@SeppoTakalo @kjbracey-arm Is the intention to get this in for 5.9?

VeijoPesonen commented:
@cmonr My changes are not fixes, refactoring of existing code, breaking change to existing code, new features and do not consider new targets. I decided to add a new category to evoke this discussion. I would say there is need to have its own category for test cases. Unless new test case is to be treated like a new feature.

Commit:
Greentea TCP test cases - check all socket calls

Commit:
Fixes Greentea UDP test cases

	udpsocket_echotest.cpp
	udpsocket_echotest_burst.cpp

VeijoPesonen commented:
Please let me know when it's fine to squash those fixup commits.

SeppoTakalo commented:
It is OK to squash when ever you feel so. And please rebase on top of master.
Start checking of the build failures.

For example:
```
Build failures:

  * K64F::ARM::TESTS-NETWORK-WIFI

        Building project wifi (K64F, ARM)

        Scan: ARM

        Scan: wifi

        Scan: FEATURE_STORAGE

        Scan: FEATURE_LWIP

        Compile [  5.3%]: get_interface.cpp

        [Error] get_interface.cpp@34,0:  #513: a value of type "EthernetInterface *" cannot be assigned to an entity of type "WiFiInterface *"
```

Once build failures are fixed, ask Kevin to start morph testing for this PR. We need to see which boards start failing.

SeppoTakalo commented:
Please remove also following tests from the tree:
* netsocket/ip_parsing <- This is unittesting the SocketAddress class. Not useful for testing HW.
* netsocket/socket_sigio <- This one is obsoleted by non-blocking test-cases
* netsocket/tcp_hello_world <- This is obsoleted by other TCP tests

SeppoTakalo commented:
Can we also refactor testcases to follow naming from the test plan?

So instead of:
```
void test_udpsocket_echotest();
...
Case cases[] = {
        Case("Echo", test_udpsocket_echotest),
```

We would use:
```
void UDPSOCKET_ECHOTEST();
...
Case cases[] = {
        Case("UDPSOCKET_ECHOTEST", UDPSOCKET_ECHOTEST),
```

That would allow us to publish the test plan without modifications.

VeijoPesonen commented:
I'll switch the naming convention. I don't see a reason to suppress compiler error for TESTS-NETWORK-WIFI test cases when Ethernet is in use. Maybe it would make more sense to use preprocessor directive explicitly to report an error, or not...

SeppoTakalo commented:
@VeijoPesonen Also, you need to do the separation of testcase.

Currently in both TCP and UDP you have
```
#ifndef MBED_EXTENDED_TESTS
#error [NOT_SUPPORTED] Pressure tests are not supported by default
#endif
```
Which is not what we are aiming for. So remove those.
If device is able to provide network interface, it should be able to run at least some of the testcases.

So do the splitting in the `Case case[]` table.
For example in `tcp/main.cpp`
```
    Case("Echo", test_tcpsocket_echotest),
    Case("Echo non-block", test_tcpsocket_echotest_nonblock),
    Case("Reuse a socket", test_tcpsocket_open_close_repeat),
    Case("Open at least 4 sockets", test_tcpsocket_open_limit),
    Case("Parallel socket thread safety", test_tcpsocket_thread_per_socket_safety),
    Case("Send repeatedly", test_tcpsocket_send_repeat),
#ifdef MBED_EXTENDED_TESTS
    Case("Echo burst", test_tcpsocket_echotest_burst),
    Case("Echo burst non-block", test_tcpsocket_echotest_burst_nonblock),
    Case("Invalid endpoint rejected", test_tcpsocket_connect_invalid),
    Case("Receive 100k from CHARGEN service", test_tcpsocket_recv_100k),
    Case("Receive 100k from CHARGEN service non-block", test_tcpsocket_recv_100k_nonblock),
    Case("Receive in given time", test_tcpsocket_recv_timeout),
    Case("Sending shall not take too long", test_tcpsocket_send_timeout),
    Case("Endpoint initiated close", test_tcpsocket_endpoint_close),
#endif
};
```

Modify the splitting so that in PR we would run only fast smoke tests that current boards are able to pass. Then later on, in nightly test those extented tests are also enable.

SeppoTakalo commented on TESTS/netsocket/tcp/tcpsocket_echotest.cpp:
```diff
@@ -0,0 +1,149 @@
+    tcpsocket_connect_to_echo_srv(sock);
+
+    static const int PKTS = 15;
+    int pkt_sizes[PKTS] = {1,2,3,4,5,6,7,8,9,10,100,10,100,1200,100};
```
Use same packet ranges as our python tests.
1,2,3,4,5,6,7,8,9,10,20,30,40,50,60,70,80,90,100,200,300,400,500,600,700,800,900,1000,1100,1200

So from 1 to 10. Then 10 to 100 in steps of 10. Then from 100 to 1200 in steps of 100.

VeijoPesonen replied:
ok, to be fixed

SeppoTakalo commented on TESTS/netsocket/tcp/tcpsocket_echotest.cpp:
```diff
@@ -0,0 +1,149 @@
+        int bytes2recv = sent;
+        while (bytes2recv) {
+            recvd = sock.recv(&(rx_buffer[sent-bytes2recv]), bytes2recv);
+            if (recvd == NSAPI_ERROR_WOULD_BLOCK) {
```
`NSAPI_ERROR_WOULD_BLOCK` should not be treated differently in blocking case. All value less than zero are errors in this case.

VeijoPesonen replied:
ok, have to take a look

SeppoTakalo commented on TESTS/netsocket/tcp/tcpsocket_echotest.cpp:
```diff
@@ -0,0 +1,149 @@
+    int sent;
+    int x = 0;
+    for (int pkt_s = pkt_sizes[x]; x < PKTS; pkt_s = pkt_sizes[x++]) {
+        fill_tx_buffer_ascii(tx_buffer, BUFF_SIZE);
```
Just a suggestion, you already have `generate_RFC_864_pattern()` that can be used for this purpose.

VeijoPesonen replied:
Good idea ~~and to be taken into use~~

Commit:
Switches Greentea netsocket test case naming convention

From lowercase to uppercase and drops 'test_'-prefix. Test case
descriptions do also match to TC names now.

Additionally all the TCs are not behing MBED_EXTENDED_TESTS
anymore.

SeppoTakalo commented on TESTS/netsocket/udp/udpsocket_echotest.cpp:
```diff
@@ -0,0 +1,154 @@
+    for (int pkt_s = pkt_sizes[s_idx]; s_idx < PKTS; pkt_s = ++s_idx) {
+        pkt_s = pkt_sizes[s_idx];
+
+        fill_tx_buffer_ascii(tx_buffer, BUFF_SIZE);
```
To be sure, clear the rx_buffer, so if we re-send same sized packet with same data, it will be shown.
`memset(rx_buffer, 0, BUFF_SIZE);

SeppoTakalo commented on TESTS/netsocket/udp/udpsocket_echotest.cpp:
```diff
@@ -0,0 +1,154 @@
+    TEST_ASSERT_EQUAL(NSAPI_ERROR_OK, sock.open(get_interface()));
+
+    static const int PKTS = 15;
+    int pkt_sizes[PKTS] = {1,2,3,4,5,6,7,8,9,10,100,10,100,1200,100};
```
As in TCP case, use same ranges.
(Perhaps update the testplan as well)

SeppoTakalo commented on TESTS/netsocket/udp/udpsocket_recv_timeout.cpp:
```diff
@@ -0,0 +1,76 @@
+        if (recvd == NSAPI_ERROR_WOULD_BLOCK) {
+            osSignalWait(SIGNAL_SIGIO, SIGIO_TIMEOUT);
+            printf("MBED: recvfrom() took: %dms\n", timer.read_ms());
+            TEST_ASSERT_INT_WITHIN(50, (100+200)/2, timer.read_ms());
```
This is weird way of saying "More than 100, but less than 200"
Why not `TEST_ASSERT_INT_WITHIN(50, 150, ...)`

SeppoTakalo commented:
Any ideas why WIFI tests now fail to build?
Did we somehow enable the build with K64F?

SeppoTakalo commented:
@VeijoPesonen 
Ah.. looks like in master it does not fail, so you have cherry-picked some changes from Juha.

We cannot merge those in yet, selectively building WiFi tests should be resolved somewhere else.

SeppoTakalo commented:
@kjbracey-arm @0xc0170 
Can you guys help, how can I tell from pre-compilation phase that
```
#if <some magic here>
#error [NOT_SUPPORTED] This is not a WiFi device
#endif
```

Other than that, test configuration now picks up interfaces by
```
interface = MBED_CONF_APP_OBJECT_CONSTRUCTION;
```

kjbracey commented:
Prior to "default network interface" stuff, you have to assume it's not a WiFi device, unless someone has specifically told you how to find a WiFi interface. Which would be the 'MBED_CONF_APP_OBJECT_CONSTRUCTION' etc the tests use?

After default network interface, I'm afraid there's still no compile-time method, but `WiFiInterface::get_default_instance()` would return NULL or not at run-time.

For Ethernet, `DEVICE_EMAC` basically does the job.

SeppoTakalo commented:
@VeijoPesonen You need then drop the changes to WiFi tests `get_interface()` function. That is not universal, and is breaking the build.

SeppoTakalo commented:
Other than those that I have now commented, all looks good.
So after those are changed, I'm OK with these changes.

juhaylinen committed:
Fix Greentea WiFi test build issues

juhaylinen commented:
How about if I add below check to to WiFi tests `get_interface()` function

```
#if !defined(MBED_CONF_APP_WIFI_SECURE_SSID) && !defined(MBED_CONF_APP_WIFI_UNSECURE_SSID)
    #error [NOT_SUPPORTED] Requires parameters from mbed_app.json
#endif
```
Or should I just revert the changes to  `get_interface()` function?

SeppoTakalo commented:
Ah, yes, true.

Add those ifdefs, as we don't have any better solution for it.

Commit:
Fixes findings from Greentea netsocket UDP test cases

Commit:
Fixes findings from Greentea netsocket TCP test cases

SeppoTakalo commented:
Looks good from my point of view.

@kjbracey-arm @0xc0170 Can we remove the "needs work" and "do not merge" labels?

@VeijoPesonen please provide log when running these tests in K64F or any other board that is able to pass all.

kjbracey reviewed (dismissed):
Test config changes are going to conflict with feature-emac - not massively, but enough to cause a bit of integration pain. Can this hold off as long as we think feature-emac may be going?

SeppoTakalo commented:
Yes, I'm OK with holding this until EMAC is merged.
EMAC is too major. This is useful even if it slips from 5.9.

0xc0170 commented:
/morph build

VeijoPesonen commented:
@SeppoTakalo the logs you were asking for.
```
mbedgt: test suite 'tests-netsocket-tcp' ............................................................. OK in 146.07 sec
        test case: 'TCPSOCKET_CONNECT_INVALID' ....................................................... OK in 18.54 sec
        test case: 'TCPSOCKET_ECHOTEST' .............................................................. OK in 1.94 sec
        test case: 'TCPSOCKET_ECHOTEST_BURST' ........................................................ OK in 26.20 sec
        test case: 'TCPSOCKET_ECHOTEST_BURST_NONBLOCK' ............................................... OK in 31.05 sec
        test case: 'TCPSOCKET_ECHOTEST_NONBLOCK' ..................................................... OK in 23.03 sec
        test case: 'TCPSOCKET_ENDPOINT_CLOSE' ........................................................ OK in 0.59 sec
        test case: 'TCPSOCKET_OPEN_CLOSE_REPEAT' ..................................................... OK in 0.06 sec
        test case: 'TCPSOCKET_OPEN_LIMIT' ............................................................ OK in 0.21 sec
        test case: 'TCPSOCKET_RECV_100K' ............................................................. OK in 3.97 sec
        test case: 'TCPSOCKET_RECV_100K_NONBLOCK' .................................................... OK in 4.00 sec
        test case: 'TCPSOCKET_RECV_TIMEOUT' .......................................................... OK in 1.17 sec
        test case: 'TCPSOCKET_SEND_REPEAT' ........................................................... OK in 8.56 sec
        test case: 'TCPSOCKET_SEND_TIMEOUT' .......................................................... OK in 0.17 sec
        test case: 'TCPSOCKET_THREAD_PER_SOCKET_SAFETY' .............................................. OK in 1.86 sec
mbedgt: test case summary: 14 passes, 0 failures
mbedgt: mbed-host-test-runner: started
mbedgt: checking for GCOV data...
mbedgt: test on hardware with target id: 07960221055369163A17FFBD
mbedgt: test suite 'tests-netsocket-udp' ............................................................. OK in 47.42 sec
        test case: 'UDPSOCKET_ECHOTEST' .............................................................. OK in 1.64 sec
        test case: 'UDPSOCKET_ECHOTEST_BURST' ........................................................ OK in 11.55 sec
        test case: 'UDPSOCKET_ECHOTEST_BURST_NONBLOCK' ............................................... OK in 6.54 sec
        test case: 'UDPSOCKET_ECHOTEST_NONBLOCK' ..................................................... OK in 2.42 sec
        test case: 'UDPSOCKET_OPEN_CLOSE_REPEAT' ..................................................... OK in 0.06 sec
        test case: 'UDPSOCKET_OPEN_LIMIT' ............................................................ OK in 0.22 sec
        test case: 'UDPSOCKET_SENDTO_REPEAT' ......................................................... OK in 0.06 sec
        test case: 'UDPSOCKET_SENDTO_TIMEOUT' ........................................................ OK in 0.09 sec
mbedgt: test case summary: 8 passes, 0 failures
mbedgt: all tests finished!
mbedgt: shuffle seed: 0.2034433001
mbedgt: test suite report:
+-------------------+---------------+---------------------+--------+--------------------+-------------+
| target            | platform_name | test suite          | result | elapsed_time (sec) | copy_method |
+-------------------+---------------+---------------------+--------+--------------------+-------------+
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-tcp | OK     | 146.07             | default     |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-udp | OK     | 47.42              | default     |
+-------------------+---------------+---------------------+--------+--------------------+-------------+
mbedgt: test suite results: 2 OK
mbedgt: test case report:
+-------------------+---------------+---------------------+------------------------------------+--------+--------+--------+--------------------+
| target            | platform_name | test suite          | test case                          | passed | failed | result | elapsed_time (sec) |
+-------------------+---------------+---------------------+------------------------------------+--------+--------+--------+--------------------+
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-tcp | TCPSOCKET_CONNECT_INVALID          | 1      | 0      | OK     | 18.54              |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-tcp | TCPSOCKET_ECHOTEST                 | 1      | 0      | OK     | 1.94               |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-tcp | TCPSOCKET_ECHOTEST_BURST           | 1      | 0      | OK     | 26.2               |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-tcp | TCPSOCKET_ECHOTEST_BURST_NONBLOCK  | 1      | 0      | OK     | 31.05              |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-tcp | TCPSOCKET_ECHOTEST_NONBLOCK        | 1      | 0      | OK     | 23.03              |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-tcp | TCPSOCKET_ENDPOINT_CLOSE           | 1      | 0      | OK     | 0.59               |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-tcp | TCPSOCKET_OPEN_CLOSE_REPEAT        | 1      | 0      | OK     | 0.06               |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-tcp | TCPSOCKET_OPEN_LIMIT               | 1      | 0      | OK     | 0.21               |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-tcp | TCPSOCKET_RECV_100K                | 1      | 0      | OK     | 3.97               |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-tcp | TCPSOCKET_RECV_100K_NONBLOCK       | 1      | 0      | OK     | 4.0                |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-tcp | TCPSOCKET_RECV_TIMEOUT             | 1      | 0      | OK     | 1.17               |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-tcp | TCPSOCKET_SEND_REPEAT              | 1      | 0      | OK     | 8.56               |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-tcp | TCPSOCKET_SEND_TIMEOUT             | 1      | 0      | OK     | 0.17               |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-tcp | TCPSOCKET_THREAD_PER_SOCKET_SAFETY | 1      | 0      | OK     | 1.86               |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-udp | UDPSOCKET_ECHOTEST                 | 1      | 0      | OK     | 1.64               |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-udp | UDPSOCKET_ECHOTEST_BURST           | 1      | 0      | OK     | 11.55              |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-udp | UDPSOCKET_ECHOTEST_BURST_NONBLOCK  | 1      | 0      | OK     | 6.54               |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-udp | UDPSOCKET_ECHOTEST_NONBLOCK        | 1      | 0      | OK     | 2.42               |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-udp | UDPSOCKET_OPEN_CLOSE_REPEAT        | 1      | 0      | OK     | 0.06               |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-udp | UDPSOCKET_OPEN_LIMIT               | 1      | 0      | OK     | 0.22               |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-udp | UDPSOCKET_SENDTO_REPEAT            | 1      | 0      | OK     | 0.06               |
| NUCLEO_F429ZI-IAR | NUCLEO_F429ZI | tests-netsocket-udp | UDPSOCKET_SENDTO_TIMEOUT           | 1      | 0      | OK     | 0.09               |
+-------------------+---------------+---------------------+------------------------------------+--------+--------+--------+--------------------+
```

0xc0170 commented:
We will restart CI, the license server issue, not related to this PR

SeppoTakalo commented:
@VeijoPesonen I'm assuming that was with `EXTENTED_TESTS` defined.
Looks good. All tests in less that 4 minutes.

VeijoPesonen commented:
@SeppoTakalo Yes, that was the case

cmonr commented:
/morph build

0xc0170 commented:
Please review the failures. repeat and timeout fail for 2 boards - something to be fixed in the targets side?

VeijoPesonen commented:
In summary:


**LPC546XX,GCC_ARM - FAILURE** - sock.sendto() returns NSAPI_ERROR_NO_MEMORY
NUCLEO_F401RE,GCC_ARM - SUCCESS
K22F,GCC_ARM - SUCCESS
KW24D,IAR - SUCCESS
**ARCH_PRO,ARM - FAILURE**
**LPC546XX,IAR - FAILURE** - sock.sendto() returns NSAPI_ERROR_NO_MEMORY
K64F,ARM - SUCCESS
NUCLEO_F429ZI,ARM - SUCCESS
NUCLEO_F401RE,IAR - SUCCESS
NUCLEO_F207ZG,ARM - SUCCESS
NRF51_DK,GCC_ARM - SUCCESS
K82F,IAR - SUCCESS
EV_COG_AD4050LZ,GCC_ARM - SUCCESS
NUCLEO_F207ZG,GCC_ARM - SUCCESS
KW24D,ARM - SUCCESS
NRF51_DK,ARM - SUCCESS
NRF52_DK,ARM - SUCCESS
KW41Z,IAR - SUCCESS
REALTEK_RTL8195AM,GCC_ARM - SUCCESS
**ARCH_PRO,IAR - FAILURE**
K82F,GCC_ARM - SUCCESS
REALTEK_RTL8195AM,ARM - SUCCESS
EV_COG_AD3029LZ,ARM - SUCCESS
NUCLEO_F401RE,ARM - SUCCESS
NUCLEO_F746ZG,GCC_ARM - SUCCESS
NUCLEO_F207ZG,IAR - SUCCESS
EV_COG_AD3029LZ,IAR - SUCCESS
KW24D,GCC_ARM - SUCCESS
**LPC546XX,ARM - FAILURE** - sock.sendto() returns NSAPI_ERROR_NO_MEMORY
K64F,IAR - SUCCESS
NRF52_DK,GCC_ARM - SUCCESS
EV_COG_AD3029LZ,GCC_ARM - SUCCESS
KW41Z,ARM - SUCCESS
MAX32630FTHR,IAR - SUCCESS
NUCLEO_F429ZI,IAR - SUCCESS
MAX32630FTHR,GCC_ARM - SUCCESS
NUCLEO_F429ZI,GCC_ARM - SUCCESS
**ARCH_PRO,GCC_ARM - FAILURE**
NUCLEO_F746ZG,IAR - SUCCESS
K64F,GCC_ARM - SUCCESS
NUCLEO_F746ZG,ARM - SUCCESS
NRF51_DK,IAR - SUCCESS
KW41Z,GCC_ARM - SUCCESS
MAX32630FTHR,ARM - SUCCESS
EV_COG_AD4050LZ,IAR - SUCCESS
REALTEK_RTL8195AM,IAR - SUCCESS
NRF52_DK,IAR - SUCCESS
K82F,ARM - SUCCESS
K22F,ARM - SUCCESS
EV_COG_AD4050LZ,ARM - SUCCESS
K22F,IAR - SUCCESS


The problematic test cases are

```
    LPC546XX-GCC_ARM.tests-netsocket-udp.UDPSOCKET_SENDTO_TIMEOUT
    ARCH_PRO-ARM.tests-netsocket-tcp.TCPSOCKET_SEND_REPEAT
    ARCH_PRO-ARM.tests-netsocket-udp.UDPSOCKET_SENDTO_REPEAT
    LPC546XX-IAR.tests-netsocket-udp.UDPSOCKET_SENDTO_TIMEOUT
    ARCH_PRO-IAR.tests-netsocket-tcp.TCPSOCKET_SEND_REPEAT
    ARCH_PRO-IAR.tests-netsocket-udp.UDPSOCKET_ECHOTEST
    LPC546XX-ARM.tests-netsocket-udp.UDPSOCKET_SENDTO_TIMEOUT
    ARCH_PRO-GCC_ARM.tests-netsocket-tcp.TCPSOCKET_SEND_REPEAT
    ARCH_PRO-GCC_ARM.tests-netsocket-udp.UDPSOCKET_SENDTO_REPEAT
```

I don't know is there much to do in out of memory situation from the test case (UDPSOCKET_SENDTO_TIMEOUT) side as it doesn't reserve memory dynamically.

UDPSOCKET_ECHOTEST, TCPSOCKET_SEND_REPEAT and UDPSOCKET_SENDTO_REPEAT are only failing(test case timeout) on ARCH_PRO so I would guess there is something fishy on the board and/or in its configuration.

VeijoPesonen commented:
@SeppoTakalo, @0xc0170  UDPSOCKET_ECHOTEST, TCPSOCKET_SEND_REPEAT and UDPSOCKET_SENDTO_REPEAT are now made as part of the extended test set. In other words those are disabled from the normal test set. Would you please review and restart the tests.

jeromecoutant commented on TESTS/netsocket/tcp/main.cpp:
```diff
@@ -0,0 +1,132 @@
+void tcpsocket_connect_to_echo_srv(TCPSocket& sock) {
+    SocketAddress tcp_addr;
+
+    get_interface()->gethostbyname(MBED_CONF_APP_ECHO_SERVER_ADDR, &tcp_addr);
```
Hi
In previous tests, there was an option to not explicitly define server address and port,
and use greentea to get dynamically the used test PC (greentea_send_kv("host_ip",...).
Maybe it can be implemented ?
Thx

SeppoTakalo replied:
The idea here is that we test against publicly available echo server, not one running on local PC.

We are no-more expecting test machine to be connected to same network as it is a limitation that cannot be done for some technologies, like Cellular for example.

jeromecoutant replied:
I agree, I don't want to modify the default public test.

Just want to add something like :

#if defined(MBED_CONF_APP_ECHO_SERVER_ADDR) && defined(MBED_CONF_APP_ECHO_SERVER_PORT)		
get_interface()->gethostbyname(MBED_CONF_APP_ECHO_SERVER_ADDR, &tcp_addr);
#else /* MBED_CONF_APP_ECHO_SERVER_ADDR && MBED_CONF_APP_ECHO_SERVER_PORT */		
/* use greentea_send_kv  and greentea_parse_kv */
get_interface()->gethostbyname(LOCAL_HOST, &tcp_addr);
#endif

SeppoTakalo replied:
If you have local echo server, you might as well provide its address in `mbed_app.json` by overriding MBED_CONF_APP_ECHO_SERVER. No need for #ifdefs.

jeromecoutant replied:
Does anyone know how to set up a local TCP/UDP echo server ?
Thx

SeppoTakalo replied:
Answered by email,

However this is very valid question.
We are preparing to release not only test environment documentation, but test case documentation as well.
There are so many expectation about environment and implementations that those cannot be guessed right by just looking at the code.

0xc0170 commented:
/morph build

Update: waiting for emac to land, we will test once this one is rebased on top of emac changes, aborting the above build

SeppoTakalo commented:
Rebased on top of master.

Emac is now in "needs Preceeding PR" label should be removed.

@0xc0170 or @kjbracey-arm please start tests for this.

kjbracey commented:
It would be nice if this could go to 5.9.0, but I think CI load may prevent it. Labelling as 5.9.1 for now.

cmonr commented:
Going to attempt to squeeze this in.

/morph build
